spec/models/user_spec.rb

```diff
@@ -0,0 +1,71 @@
+require "rails_helper"
+require "shared/share_example_spec.rb"
+
+RSpec.describe User, type: :model do
+  subject {FactoryBot.create(:user)} 
+
+  it "has a valid factory" do
+    expect(subject).to be_valid
+  end
+
+  describe "associations" do
+    it {is_expected.to have_many(:orders).dependent(:destroy)}
+    it {is_expected.to have_many(:addresses).dependent(:destroy)}
+  end
+
+  describe "#email_downcase" do
+    let(:user){FactoryBot.create :user, email: "EXAMPLE@gmail.com"}
+
+    it "return email downcase" do
+      user.send :email_downcase
+      expect(user.email).to eq("example@gmail.com")
```

Comment:
anh nhớ là cái này tự downcase khi em create new record mà, line 20 hơi thừa nhỉ

Reply:
Dạ cái này em gọi đến method email_downcase ạ, không có dòng này chạy nó bị lỗi á a nạ :v

Reply:
em bjk sao lỗi ko
1. vì method downcase nó dc chạy khi em create hay update email ấy
nên chỗ này em nên tạo xử lý như tạo mới user hay update user với đầu vào em truyền vào là email dc  in hoa, và sau khi lưu vào DB thì lại đc downcase rồi

2. sẵn đây, nhắc nhẹ method send sẽ giúp em gọi đến method bên trong model user, nhưng nó gọi dc tất, tức là public, hay protected hay private cũng gọi dc, send nó đang vi phạm tính chất đóng gói của OOP nhé. thường thì ng ta tránh việc đó bằng cách dùng public method

chị giiar thích thêm cho em hiểu thôi

chỗ nào em đổi lại xử lý để test cho cái calll back before_save nhé